stan/math/prim/fun/generalized_inverse.hpp

```diff
@@ -31,15 +31,22 @@ template <typename EigMat, require_eigen_t<EigMat>* = nullptr,
 inline Eigen::Matrix<value_type_t<EigMat>, EigMat::ColsAtCompileTime,
                      EigMat::RowsAtCompileTime>
 generalized_inverse(const EigMat& G) {
-  if (G.size() == 0)
+  const auto& G_ref = to_ref(G);
+  if (G_ref.size() == 0)
     return {};
 
-  if (G.rows() == G.cols())
-    return inverse(G);
-
-  const auto& G_ref = to_ref(G);
+  if (G_ref.rows() == G_ref.cols()) {
+    Eigen::CompleteOrthogonalDecomposition<
+        Eigen::Matrix<value_type_t<EigMat>, EigMat::RowsAtCompileTime,
+                      EigMat::ColsAtCompileTime>>
+        complete_ortho_decomp_G = G_ref.completeOrthogonalDecomposition();
+    if (!(complete_ortho_decomp_G.rank() < G_ref.rows()))
+      return inverse(G_ref);
```

Comment:
So my only Q is how much this costs / accuracy of computing the `completeOrthogonalDecomposition()` and then using it for the check and then sometimes using it for the pseudoinverse. If someone passes an NxM matrix here i sort of feels like they either

1. Would know ahead of time that they will or will not have a square matrix
2. Know their matrix is low rank and they can't use `inverse()`

Are those not good assumptions? If they are then should we always just do the pseudoinverse()? Is the accuracy of pseudoinverse pretty low compared to inverse()? I almost feel like even if so someone using this would expect a pseudoinverse instead of the inverse

Reply:
These are great questions!

On
> If someone passes an NxM matrix here i sort of feels like they either
> 
>   1. Would know ahead of time that they will or will not have a square matrix
>   2. Know their matrix is low rank and they can't use `inverse()`

I'm interested in the case where you'd specifically pass a square matrix but not know ahead of time if the matrix is low rank or not. It's probably low-rank but it may not be (say in a factor model).

This got me to revisit the paper that the code is based on. The author does something that I missed because he notates `L` as a cholesky factor when it is actually a modified cholesky factor that works even for low rank matrices. This is how he is able to not do a rank find (which one usually does either as I've done here or using SVD). I've coded up the prim version at https://github.com/spinkney/math/tree/generalized_inverse_low_rank_chol. The point is that some rank issue must be dealt with or there's numerical instability issues.

The code now follows the [paper](https://arxiv.org/ftp/arxiv/papers/0804/0804.4809.pdf) explicitly. The major change involves adding a ` cholesky_low_rank_decomposition` lambda. I've coded it using Eigen block notation. I believe there's further optimizations that could be done, if someone with a background in high performance computing helps. It could be added to Stan as a separate function, which is cool in it's own right. 

The prim tests pass. I haven't benchmarked though to see how it compares to the `completeOrthogonalDecomposition()` version here.

Reply:
I've put together an example how low rank chol can transform a low rank matrix of "factors" into a valid correlation matrix. I'm not sure if this works all the time but it's promising. [low_rank_chol_eg.R.zip](https://github.com/stan-dev/math/files/7147756/low_rank_chol_eg.R.zip)

Reply:
Oh nice! So with the low rank chol version should we close this PR and open one up based on the low rank chol? Or should we merge this then make a separate PR for that?

Reply:
I'm thinking that the low rank chol should be the final version. Though I think it may be worth just getting this update in for the next release and then doing 2 PRs: 1) add low rank chol as a function and 2) update gen inverse using it in a new PR.